busted.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.16.
.TH BUSTED "1" "August 2022" "busted 2.1.1" "User Commands"
.SH NAME
busted \- Lua unit testing framework focused on ease of use
.SH DESCRIPTION
busted: error: Usage: busted [OPTIONS] [\-\-] [ROOT\-1 [ROOT\-2 [...]]]
.SS "ARGUMENTS:"
.TP
ROOT
test script file/folder. Folders will be
traversed for any file that matches the
\fB\-\-pattern\fR option. (optional, default: nil)
.SS "OPTIONS:"
.TP
\fB\-\-version\fR
prints the program version and exits
.TP
\fB\-p\fR, \fB\-\-pattern\fR=\fI\,PATTERN\/\fR
only run test files matching the Lua
pattern (default: _spec)
.TP
\fB\-\-exclude\-pattern\fR=\fI\,PATTERN\/\fR do not run test files matching the Lua
pattern, takes precedence over \fB\-\-pattern\fR
.TP
\fB\-e\fR STATEMENT
execute statement STATEMENT
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,LIBRARY\/\fR
output library to load (default:
plainTerminal)
.TP
\fB\-C\fR, \fB\-\-directory\fR=\fI\,DIR\/\fR
change to directory DIR before running
tests. If multiple options are specified,
each is interpreted relative to the
previous one. (default: ./)
.TP
\fB\-\-coverage\-config\-file\fR=\fI\,FILE\/\fR
load luacov configuration options from FILE
.TP
\fB\-f\fR, \fB\-\-config\-file\fR=\fI\,FILE\/\fR
load configuration options from FILE
.TP
\fB\-t\fR, \fB\-\-tags\fR=\fI\,TAGS\/\fR
only run tests with these #tags (default:
[])
.TP
\fB\-\-exclude\-tags\fR=\fI\,TAGS\/\fR
do not run tests with these #tags, takes
precedence over \fB\-\-tags\fR (default: [])
.TP
\fB\-\-filter\fR=\fI\,PATTERN\/\fR
only run test names matching the Lua
pattern (default: [])
.TP
\fB\-\-filter\-out\fR=\fI\,PATTERN\/\fR
do not run test names matching the Lua
pattern, takes precedence over \fB\-\-filter\fR
(default: [])
.TP
\fB\-m\fR, \fB\-\-lpath\fR=\fI\,PATH\/\fR
optional path to be prefixed to the Lua
module search path (default:
\&./src/?.lua;./src/?/?.lua;./src/?/init.lua)
.TP
\fB\-\-cpath\fR=\fI\,PATH\/\fR
optional path to be prefixed to the Lua C
module search path (default:
\&./csrc/?.so;./csrc/?/?.so;)
.TP
\fB\-r\fR, \fB\-\-run\fR=\fI\,RUN\/\fR
config to run from .busted file
.TP
\fB\-\-repeat\fR=\fI\,COUNT\/\fR
run the tests repeatedly (default: 1)
.TP
\fB\-\-seed\fR=\fI\,SEED\/\fR
random seed value to use for shuffling test
order (default: \fI\,/dev/urandom\/\fP or os.time())
.TP
\fB\-\-lang\fR=\fI\,LANG\/\fR
language for error messages (default: en)
.TP
\fB\-\-loaders\fR=\fI\,NAME\/\fR
test file loaders (default: lua,moonscript)
.TP
\fB\-\-helper\fR=\fI\,PATH\/\fR
A helper script that is run before tests
.TP
\fB\-\-lua\fR=\fI\,LUA\/\fR
The path to the lua interpreter busted
should run under
.TP
\fB\-Xoutput\fR OPTION
pass `OPTION` as an option to the output
handler. If `OPTION` contains commas, it is
split into multiple options at the commas.
(default: [])
.TP
\fB\-Xhelper\fR OPTION
pass `OPTION` as an option to the helper
script. If `OPTION` contains commas, it is
split into multiple options at the commas.
(default: [])
.TP
\fB\-c\fR, \fB\-\-[no\-]coverage\fR
do code coverage analysis (requires
`LuaCov` to be installed) (default: off)
.TP
\fB\-v\fR, \fB\-\-[no\-]verbose\fR
verbose output of errors (default: off)
.TP
\fB\-s\fR, \fB\-\-[no\-]enable\-sound\fR
executes `say` command if available
(default: off)
.TP
\fB\-l\fR, \fB\-\-list\fR
list the names of all tests instead of
running them
.TP
\fB\-\-ignore\-lua\fR
Whether or not to ignore the lua directive
.TP
\fB\-\-[no\-]lazy\fR
use lazy setup/teardown as the default
(default: off)
.TP
\fB\-\-[no\-]auto\-insulate\fR
enable file insulation (default: on)
.TP
\fB\-k\fR, \fB\-\-[no\-]keep\-going\fR
continue as much as possible after an error
or failure (default: on)
.TP
\fB\-R\fR, \fB\-\-[no\-]recursive\fR
recurse into subdirectories (default: on)
.TP
\fB\-\-[no\-]shuffle\fR
randomize file and test order, takes
precedence over \fB\-\-sort\fR (\fB\-\-shuffle\-test\fR and
\fB\-\-shuffle\-files\fR) (default: off)
.TP
\fB\-\-[no\-]shuffle\-files\fR
randomize file execution order, takes
precedence over \fB\-\-sort\-files\fR (default: off)
.TP
\fB\-\-[no\-]shuffle\-tests\fR
randomize test order within a file, takes
precedence over \fB\-\-sort\-tests\fR (default: off)
.TP
\fB\-\-[no\-]sort\fR
sort file and test order (\fB\-\-sort\-tests\fR and
\fB\-\-sort\-files\fR) (default: off)
.TP
\fB\-\-[no\-]sort\-files\fR
sort file execution order (default: off)
.TP
\fB\-\-[no\-]sort\-tests\fR
sort test order within a file (default:
off)
.TP
\fB\-\-[no\-]suppress\-pending\fR
suppress `pending` test output (default:
off)
.TP
\fB\-\-[no\-]defer\-print\fR
defer print to when test suite is complete
(default: off)
.PP
; re\-run with \fB\-\-help\fR for usage.
.SH AUTHOR
 This manpage was written by Nilesh Patra for the Debian distribution and
 can be used for any other usage of the program.


src/busted.lua
require'pl' -- for pretty.write table formating
assert = require'luassert'
spy = require('luassert.spy')
mock = require('luassert.mock')
stub = require('luassert.stub')

busted = {}
busted._COPYRIGHT   = "Copyright (c) 2012 Olivine Labs, LLC."
busted._DESCRIPTION = "A unit testing framework with a focus on being easy to use."
busted._VERSION     = "Busted 1.4"

-- Load default language pack
require('busted.languages.en')

local assert_proxy_call = getmetatable(assert.is_truthy).__call
local assert_call = getmetatable(assert).__call
local globals = _G
local push = table.insert
local tests = {}
local done = {}
local started = {}
local test_index = 1
local options

local step = function(...)
   local steps = {...}
   if #steps == 1 and type(steps[1]) == 'table' then
      steps = steps[1]
   end
   local i = 0
   local next
   next = function()
      i = i + 1
      local step = steps[i]
      if step then
         step(next)
      end
   end
   next()
end

busted.step = step

local next_test
next_test = function()
   if #done == #tests then
      return
   end
   if not started[test_index] then
      started[test_index] = true
      local test = tests[test_index]
      local steps = {}
      local execute_test = function(next)
         test.status = {
            description = test.name,
            info = test.info,
            trace = ''
         }         
         -- this part is a bit nasty!
         -- intercept all calls to luassert states / proxies.
         -- uses much of internal knowlage of luassert!!!!
         -- the metatable of is_truthy is the same as for other
         -- luasserts.
         getmetatable(assert.is_truthy).__call = function(...)
            local results = {pcall(assert_proxy_call,...)}
            local args = {...}
            local is_proxy = true
            -- ducktype if this is an assertion 'result' and not a proxy
            for k,v in pairs(args[1] or {}) do
               if k == 'positive_message' or k == 'negative_message' then
                  is_proxy = false
               end
            end
            if is_proxy then
               return unpack(results,2)
            else
               if results[1] and not test.status.type then
                  test.status.type = 'success'
               elseif not results[1] and test.status.type ~= 'failure' then
                  -- the error message and traceback are always caused
                  -- by the first failed assertion
                  test.status.trace = debug.traceback("", 2)
                  test.status.type = 'failure'
                  test.status.err = results[2]
                  -- dont call done() here but continue test execution
                  -- uncaught errors following, will be catched in pcall(test.f,done)
               end
            end
         end
         getmetatable(assert).__call = function(...)
            local results = {pcall(assert_call,...)}
            if results[1] and not test.status.type then
               test.status.type = 'success'
            elseif not results[1] and test.status.type ~= 'failure' then
               test.status.trace = debug.traceback("", 2)
               test.status.type = 'failure'
               test.status.err = results[2]
            end
         end
         local done = function()
            done[test_index] = true
            if test.status.type ~= 'success' and not test.status.err then
               test.status.type = 'pending'
            end
            if not options.debug and not options.defer_print then
               options.output.currently_executing(test.status, options)
            end
            test.context:decrement_test_count()
            next()
         end
         local ok,err = pcall(test.f,done)
         if not ok then
            if not test.status.err or test.status.type ~= 'failure' then
               if type(err) == "table" then
                  err = pretty.write(err)
               end
               test.status.type = 'failure'
               test.status.trace = debug.traceback("", 2)
               test.status.err = err
            end
            done()
         end
      end

      local check_before = function(context)
         if context.before then
            local execute_before = function(next)
               context.before(
                  function()
                     context.before = nil
                     next()
                  end)
            end
            push(steps,execute_before)
         end
      end

      local parents = test.context.parents

      for p=1,#parents do
         check_before(parents[p])
      end

      check_before(test.context)

      for p=1,#parents do
         if parents[p].before_each then
            push(steps,parents[p].before_each)
         end
      end

      if test.context.before_each then
         push(steps,test.context.before_each)
      end

      push(steps,execute_test)

      if test.context.after_each then
         push(steps,test.context.after_each)
      end

      local post_test = function(next)
         local post_steps = {}
         local check_after = function(context)
            if context.after then
               if context:all_tests_done() then
                  local execute_after = function(next)
                     context.after(
                        function()
                           context.after = nil
                           next()
                        end)
                  end
                  push(post_steps,execute_after)
               end
            end
         end

         check_after(test.context)

         for p=#parents,1,-1 do
            if parents[p].after_each then
               push(post_steps,parents[p].after_each)
            end
         end

         for p=#parents,1,-1 do
            check_after(parents[p])
         end

         local forward = function(next)
            test_index = test_index + 1
            next_test()
            next()
         end
         push(post_steps,forward)
         step(post_steps)
      end
      push(steps,post_test)
      step(steps)
   end
end

local create_context = function(desc)
   local context = {
      desc = desc,
      parents = {},
      test_count = 0,
      increment_test_count = function(self)
         self.test_count = self.test_count + 1
         for _,parent in ipairs(self.parents) do
            parent.test_count = parent.test_count + 1
         end
      end,
      decrement_test_count = function(self)
         self.test_count = self.test_count - 1
         for _,parent in ipairs(self.parents) do
            parent.test_count = parent.test_count - 1
         end
      end,
      all_tests_done = function(self)
         return self.test_count == 0
      end,
      add_parent = function(self,parent)
         push(self.parents,parent)
      end
   }
   return context
end

local suite_name
local current_context
busted.describe = function(desc,more)
   if not suite_name then
      suite_name = desc
   end
   local context = create_context(desc)
   for i,parent in ipairs(current_context.parents) do
      context:add_parent(parent)
   end
   context:add_parent(current_context)
   local old_context = current_context
   current_context = context
   more()
   current_context = old_context
end

busted.before = function(sync_before,async_before)
   if async_before then
      current_context.before = async_before
   else
      current_context.before = function(done)
         sync_before()
         done()
      end
   end
end

busted.before_each = function(sync_before,async_before)
   if async_before then
      current_context.before_each = async_before
   else
      current_context.before_each = function(done)
         sync_before()
         done()
      end
   end
end

busted.after = function(sync_after,async_after)
   if async_after then
      current_context.after = async_after
   else
      current_context.after = function(done)
         sync_after()
         done()
      end
   end
end

busted.after_each = function(sync_after,async_after)
   if async_after then
      current_context.after_each = async_after
   else
      current_context.after_each = function(done)
         sync_after()
         done()
      end
   end
end

busted.pending = function(name)
   local test = {}
   test.context = current_context
   test.context:increment_test_count()
   test.name = name
   local debug_info = debug.getinfo(2)
   test.f = function(done)
      done()
   end
   test.info = {
      source = debug_info.source,
      short_src = debug_info.short_src,
      linedefined = debug_info.linedefined,
   }
   tests[#tests+1] = test
end

busted.it = function(name,sync_test,async_test)
   local test = {}
   test.context = current_context
   test.context:increment_test_count()
   test.name = name
   local debug_info
   if async_test then
      debug_info = debug.getinfo(async_test)
      test.f = async_test
   else
      debug_info = debug.getinfo(sync_test)
      -- make sync test run async
      test.f = function(done)
         sync_test()
         done()
      end
   end
   test.info = {
      source = debug_info.source,
      short_src = debug_info.short_src,
      linedefined = debug_info.linedefined,
   }
   tests[#tests+1] = test
end

busted.reset = function()
   current_context = create_context('Root context')
   tests = {}
   done = {}
   started = {}
   test_index = 1
   suite_name = nil
end

local play_sound = function(failures)
   math.randomseed(os.time())
   
   if options.failure_messages and #options.failure_messages > 0 and
      options.success_messages and #options.success_messages > 0 then
      if failures and failures > 0 then
         io.popen("say \""..options.failure_messages[math.random(1, #options.failure_messages)]:format(failures).."\"")
      else
         io.popen("say \""..options.success_messages[math.random(1, #options.success_messages)].."\"")
      end
   end
end

busted.run = function(opts)
   options = opts
   local ms = os.clock()
   
   suite_name = suite_name or 'Root context'

   if not options.debug and not options.defer_print then      
      print(options.output.header(suit_name,#tests))
   end

   local loop = options.loop or function() end

   repeat
      next_test()
      loop()
   until #done == #tests

   ms = os.clock() - ms

   if not options.debug and options.defer_print then
      print(options.output.header(suit_name,#tests))
   end

   local statuses = {}
   local failures = 0
   for _,test in ipairs(tests) do
      push(statuses,test.status)
      if test.status.type == 'failure' then
         failures = failures + 1
      end
   end
   
   if options.sound then
      play_sound(failures)
   end
   if options.debug then
      return statuses
   else
      if options.defer_print then
         print(options.output.footer(failures))
      end
      return options.output.formatted_status(statuses, options, ms), failures
   end
end

it = busted.it
pending = busted.pending
describe = busted.describe
before = busted.before
after = busted.after
setup = busted.before
teardown = busted.after
before_each = busted.before_each
after_each = busted.after_each
step = step

-- only for internal testing
busted.setup_async_tests = function(yield,loopname)
   describe(
      loopname..' test suite',
      function()
         local before_each_count = 0
         local before_called
         before(
            async,
            function(done)
               yield(
                  function()
                     before_called = true
                     done()
                  end)
            end)

         before_each(
            async,
            function(done)
               yield(
                  function()
                     before_each_count = before_each_count + 1
                     done()
                  end)
            end)

         it(
            'order 1 should async succeed',
            async,
            function(done)
               yield(
                  function()
                     assert.is_true(before_called)
                     assert.is.equal(before_each_count,1)
                     done()
                  end)
            end)

         it(
            'order 2 should async fail',
            async,
            function(done)
               yield(
                  function()
                     assert.is_truthy(false)
                     done()
                  end)
            end)

         it(
            'order 3 should async fails epicly',
            async,
            function(done)
               does_not_exist.foo = 3
            end)

         it(
            'order 4 should async have no assertions and pends thus',
            async,
            function(done)
               done()
            end)

         it(
            'order 5 spies should sync succeed',
            function()
               assert.is.equal(before_each_count,5)
               local thing = {
                  greet = function()
                  end
               }
               spy.on(thing, "greet")
               thing.greet("Hi!")
               assert.spy(thing.greet).was.called()
               assert.spy(thing.greet).was.called_with("Hi!")
            end)

         it(
            'order 6 spies should async succeed',
            async,
            function(done)
               local thing = {
                  greet = function()
                  end
               }
               spy.on(thing, "greet")
               yield(
                  function()
                     assert.spy(thing.greet).was.called()
                     assert.spy(thing.greet).was.called_with("Hi!")
                     done()
                  end)
               thing.greet("Hi!")
            end)

         describe(
            'with nested contexts',
            function()
               local before_called
               before(
                  async,
                  function(done)
                     yield(
                        function()
                           before_called = true
                           done()
                        end)
                  end)
               it(
                  'order 7 nested async test before is called succeeds',
                  async,
                  function(done)
                     yield(
                        function()
                           assert.is_true(before_called)
                           done()
                        end)
                  end)
            end)
         
         pending('order 8 is pending')
      end)
end

-- only for internal testing
busted.describe_statuses = function(statuses,print_statuses)
   if print_statuses then
      print('---------- STATUSES ----------')
      print(pretty.write(statuses))
      print('------------------------------')
   end

   describe(
      'Test statuses',
      function()
         it(
            'execution order is correct',
            function()
               for i,status in ipairs(statuses) do
                  local order = status.description:match('order (%d+)')
                  assert.is.equal(tonumber(order),i)
               end
            end)

         it(
            'type is correct',
            function()
               for i,status in ipairs(statuses) do
                  local type = status.type
                  assert.is_truthy(type == 'failure' or type == 'success' or type == 'pending')
                  local succeed = status.description:match('succeed')
                  local fail = status.description:match('fail')
                  local pend = status.description:match('pend')
                  local count = 0
                  if succeed then
                     count = count + 1
                  end
                  if fail then
                     count = count + 1
                  end
                  if pend then
                     count = count + 1
                  end
                  assert.equal(count,1)
                  if succeed then
                     assert.is.equal(status.type,'success')
                  elseif fail then
                     assert.is.equal(status.type,'failure')                  
                  elseif pend then
                     assert.is.equal(status.type,'pending')
                  end
               end
            end)

         it(
            'info is correct',
            function()
               for i,status in ipairs(statuses) do
                  --  print(pretty.write(status))
                  assert.is_truthy(status.info.linedefined)
                  assert.is_truthy(status.info.source:match('busted%.lua'))
                  assert.is_truthy(status.info.short_src:match('busted%.lua'))
               end
            end)

         it(
            'provides "err" for failed tests',
            function()
               for i,status in ipairs(statuses) do
                  if status.type == 'failure' then
                     assert.is.equal(type(status.err),'string')
                     assert.is_not.equal(#status.err,0)
                  end
               end
            end)

         it(
            'provides "traceback" for failed tests',
            function()
               for i,status in ipairs(statuses) do
                  if status.type == 'failure' then
                     assert.is.equal(type(status.trace),'string')
                     assert.is_not.equal(#status.trace,0)
                  end
               end
            end)

      end)
end

return busted